File: js/problem27.js
/**
 * @param {number[]} nums
 * @param {number} val
 * @return {number}
 */
var removeElement = function(nums, val) {
    if (nums.length === 0) {
        return 0;
    }
    
    let i = 0, j = 0, res = 0;
    
    while (j < nums.length) {
        if (nums[j] === val) {
            j++;
            if (j < nums.length) {
                nums[i] = nums[j];
            }
            res++;
        } else {
            nums[i] = nums[j];
            i++;
            j++;
        }
    }
    
    return nums.length - res;
};

File: leetcodeJava/src/main/java/Problem27.java
// 27. Remove Element

class Solution {
    public int removeElement(int[] nums, int val) {
        if (nums.length == 0) {
            return 0;
        }
        
        int i = 0;
        int j = nums.length - 1;
        
        while (i < j) {
            if (nums[i] == val) {
                int temp = nums[i];
                nums[i] = nums[j];
                nums[j] = temp;
                j--;
            } else {
                i++;
            }
        }
        
        if (nums[i] == val) {
            return i;
        }
        
        return i + 1;
    }
}